File: Misc/131_lc_1466.py
'''
1466. Reorder Routes to Make All Paths Lead to the City Zero
'''

from typing import List
from collections import deque, defaultdict

class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        directedGraph = defaultdict(list)
        undirectedGraph = defaultdict(list)
        for fromCity, toCity in connections:
            directedGraph[fromCity].append(toCity)

            undirectedGraph[fromCity].append(toCity)
            undirectedGraph[toCity].append(fromCity)

        res = 0
        q = deque()
        q.append(0)
        visited = set()
        visited.add(0)

        while q:
            qSize = len(q)
            for i in range(qSize):
                fromCity = q.popleft()
                for toCity in undirectedGraph.get(fromCity, []):
                    if toCity not in visited:
                        if fromCity not in directedGraph.get(toCity, []):
                            res+=1
                        q.append(toCity)
                        visited.add(toCity)
                    
        
        return res

File: Misc/131_lc_1466.java
/*
 * 1466. Reorder Routes to Make All Paths Lead to the City Zero
 */

import java.util.*;

class Solution {
    public int minReorder(int n, int[][] connections) {
        Map<Integer, List<Integer>> directedGraph = new HashMap<>();
        Map<Integer, List<Integer>> undirectedGraph = new HashMap<>();
        
        // Build the directed and undirected graphs
        for (int[] connection : connections) {
            int fromCity = connection[0];
            int toCity = connection[1];
            directedGraph.putIfAbsent(fromCity, new ArrayList<>());
            directedGraph.get(fromCity).add(toCity);
            
            undirectedGraph.putIfAbsent(fromCity, new ArrayList<>());
            undirectedGraph.get(fromCity).add(toCity);
            undirectedGraph.putIfAbsent(toCity, new ArrayList<>());
            undirectedGraph.get(toCity).add(fromCity);
        }

        int res = 0;
        Queue<Integer> queue = new LinkedList<>();
        queue.add(0);
        Set<Integer> visited = new HashSet<>();
        visited.add(0);

        // BFS traversal
        while (!queue.isEmpty()) {
            int fromCity = queue.poll();
            for (int toCity : undirectedGraph.getOrDefault(fromCity, new ArrayList<>())) {
                if (!visited.contains(toCity)) {
                    if (!directedGraph.getOrDefault(toCity, new ArrayList<>()).contains(fromCity)) {
                        res++; // Need to reorder this road
                    }
                    queue.add(toCity);
                    visited.add(toCity);
                }
            }
        }

        return res;
    }
}
